Fix Pandora Presence

FireController1847 opened the pull request:
There have been many a' update since the Pandora presence was created, from the logo having changed to the way data is handled. In particular, I have Pandora Premium and noticed the presence was very incompatible with it (presuming the original creator did not have premium, so they couldn't test with it).

Previously, the presence would fetch an audio source off of the webpage and use that to fetch various pieces of information. The problem is, Pandora now obfuscates, encodes, and streams audio sources differently if you're not listening from a station but a playlist. So, I removed the dependency on the audio source entirely and instead made it entirely rely on the audio bar, of which the title and artist were already being fetched.

I love Pandora so I'm glad I was able to contribute to this, hopefully I rewrote it to an acceptable standard. Let me know if there's anything I can change.

FireController1847 committed:
Fix Pandora

EncryptedDev reviewed (changes requested):
Resolve ESlint issues

FireController1847 committed:
Version Bump

FireController1847 committed:
Fix ESLint Issues

FireController1847 committed:
Am I stupid? What just happened?

FireController1847 committed:
Remove trailing whitespaces

Kinda salty. They weren't there until I ran it through ESLint, it put them there and then framed me saying it was my fault lol

EncryptedDev commented on websites/P/Pandora/presence.ts:
```diff
@@ -1,85 +1,89 @@
-  }
+  // Set default data
+  data.details = "Browsing...";
+  data.largeImageKey = "pandora";
```
Set this in the PresenceData obj

FireController1847 replied:
Good point!

EncryptedDev commented on websites/P/Pandora/presence.ts:
```diff
@@ -1,85 +1,89 @@
+   audioBar: HTMLElement = document.querySelector(".Tuner__Audio__NowPlayingHitArea");
+  
+  // If the audio bar exists, assume we're listening to something
+  if (audioBar !== null) {
```
```suggestion
  if (audioBar) {
```

EncryptedDev commented on websites/P/Pandora/presence.ts:
```diff
@@ -1,85 +1,89 @@
+  data.largeImageKey = "pandora";
+
+  // Define whether or not we're currently playing
+  const isPlaying = true,
```
You have another constant called the same on L65?

FireController1847 replied:
Whoops, meant to set the variable not define a new one

EncryptedDev commented on websites/P/Pandora/presence.ts:
```diff
@@ -1,85 +1,89 @@
+     artist: HTMLElement = document.querySelector(".Tuner__Audio__TrackDetail__artist");
+
+    // Return if either of them are null
+    if (title === null || artist === null) 
+      return;
```
What

FireController1847 replied:
I honestly don't know. This is how it was coded before, I had assumed it had some importance. I'll change the logic behind null checking here.

EncryptedDev commented on websites/P/Pandora/presence.ts:
```diff
@@ -1,85 +1,89 @@
-    data.smallImageKey = "pause";
-    data.smallImageText = (await strings).pause;
+    // Return if either are null
+    if (timeElapsed === null || timeRemaining === null) 
+      return;
```
What

EncryptedDev commented on websites/P/Pandora/presence.ts:
```diff
@@ -1,85 +1,89 @@
+    const playButton: HTMLElement = document.querySelector(".Tuner__Control__Play__Button");
+
+    // Return if null
+    if (playButton === null) 
+      return;
```
What

EncryptedDev commented on websites/P/Pandora/presence.ts:
```diff
@@ -1,85 +1,89 @@
+      data.smallImageText = (await strings).play;
+
+      // Get timestamps
+      const [startTime, endTime] = presence.getTimestamps(
```
```suggestion
      [data.startTimestamp, data.endTimestamp] = presence.getTimestamps(
```

EncryptedDev commented on websites/P/Pandora/presence.ts:
```diff
@@ -1,85 +1,89 @@
+        presence.timestampFromFormat(stripText(timeRemaining, "Time Remaining"))
+      );
+
+      // Set timestamps
+      data.startTimestamp = startTime;
+      data.endTimestamp = endTime;
```
```suggestion
```

FireController1847 committed:
Make requested changes

FireController1847 committed:
Remove trailing whitespaces (for real this time)

Slowlife01 commented on websites/P/Pandora/presence.ts:
```diff
@@ -1,85 +1,87 @@
-      presence.info(
-        `%cPandora%cERROR%c An error occurred while stripping data off the page. Please contact Alanexei on the PreMiD Discord server, and send him a screenshot of this error. ID: ${id}`
+      presence.error(
+        `An error occurred while stripping data off the page. Please contact Alanexei on the PreMiD Discord server, and send him a screenshot of this error. ID: ${id}`
```
Alanexei is not in PreMiD server anymore, so just either tell them to create an issue on github or contact you

FireController1847 replied:
Okay, I'd be more than willing to maintain this one so I'll change it to me

Slowlife01 commented on websites/P/Pandora/presence.ts:
```diff
@@ -1,85 +1,87 @@
+            presence.timestampFromFormat(stripText(timeElapsed, "Time Elapsed")),
+            presence.timestampFromFormat(stripText(timeRemaining, "Time Remaining"))
+          );
+        } else
+          presence.error("Timestamps are null!");
```
```suggestion
        } else presence.error("Timestamps are null!");
```

Slowlife01 commented on websites/P/Pandora/presence.ts:
```diff
@@ -1,85 +1,87 @@
+        data.smallImageKey = "pause";
+        data.smallImageText = (await strings).pause;
+      }
+    } else
+      presence.error("Play button is null!");
```
```suggestion
    } else presence.error("Play button is null!");
```

Slowlife01 commented on websites/P/Pandora/presence.ts:
```diff
@@ -1,85 +1,87 @@
   }
 
-  presence.setActivity(data, audioElement ? !audioElement.paused : true);
+  presence.setActivity(data, isPlaying);
```
```suggestion
  presence.setActivity(data);
```

Slowlife01 commented on websites/P/Pandora/presence.ts:
```diff
@@ -1,85 +1,87 @@
-    ".Tuner__Audio__NowPlayingHitArea"
-  );
+  // Define whether or not we're currently playing
+  let isPlaying = true;
```
Why should this be true by default?

FireController1847 replied:
Not sure, I was keeping the old functionality. Will set to false.

Slowlife01 commented on websites/P/Pandora/presence.ts:
```diff
@@ -1,85 +1,87 @@
+      // Set them to the presence
+      data.details = stripText(title, "Title");
+      data.state = stripText(artist, "Artist");
+    } else
+      presence.error("Title and artist are null!");
```
```suggestion
    } else presence.error("Title and artist are null!");
```

FireController1847 committed:
Apply suggestions from code review

Co-authored-by: Slowlife <slowlife1165@gmail.com>

Slowlife01 commented on websites/P/Pandora/dist/metadata.json:
```diff
@@ -8,6 +8,10 @@
       "id": "163319338403627008"
+    },
+    {
+      "name": "FireController#1847",
```
```suggestion
      "name": "FireController1847",
```

FireController1847 committed:
Make requested changes

Slowlife01 commented:
Upload proof of it working

FireController1847 commented:
> Upload proof of it working

[![Proof of it working]](https://user-images.githubusercontent.com/16494272/141931912-15487263-5af9-4f07-bba0-5ea395e4aa31.mp4)

Slowlife01 committed:
style(*)

Slowlife01 reviewed (approved):
lgtm
